### PR TITLE
Ensure initializedContextIds_ map is cleaned up appropriately in DistEngine.

### LINKED ISSUE #29083: initializedContextIds_ field in DistEngine grows without bound
## 🐛 Bug

`initializedContextIds_` field in DistEngine grows without bound

## Expected behavior

We should cleanup entries for this map as autograd context ids are cleaned up. Or we should have a field in `DistAutogradContext` instead.

## Additional context

https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/engine/dist_engine.h#L86

cc @ezyang @gchanan @zou3519 @jerryzh168 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @rohan-varma @xush6528

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29787 Ensure initializedContextIds_ map is cleaned up appropriately in DistEngine.**

The initializedContextIds_ map was never cleaned up in DistEngine and
kept on growing as we continue to run backward passes. To fix this, in this PR
we ensure that the context id is cleaned up from this map once we are done with
the backward pass.

Closes #29083

Differential Revision: [D18498937](https://our.internmc.facebook.com/intern/diff/D18498937/)

### CODE REVIEW

File: torch/csrc/distributed/autograd/engine/dist_engine.h [hunk 1]
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <gtest/gtest_prod.h>
```
Comment (pietern) [inline torch/csrc/distributed/autograd/engine/dist_engine.h]: I don't like this... I know it's for making protected functions testable, but isn't there another way to achieve the same? For example, do the functions you're calling into need to be protected?

**Review by pietern — COMMENTED** [review]
Not a fan of including the gtest header... 

Otherwise the change LGTM.

File: torch/csrc/distributed/autograd/engine/dist_engine.cpp [hunk 2]
```diff
@@ -212,6 +212,7 @@ void DistEngine::executeSendFunction(
 
     // Mark the autograd context id as initialized and unlock.
     initializedContextIds_.insert(autogradContext.contextId());
+    ClearContextIdGuard guard(autogradContext.contextId());
```
Comment (mrshenli) [inline torch/csrc/distributed/autograd/engine/dist_engine.cpp]: why this needs to be before the unlock?

**Review by mrshenli — APPROVED** [review]
This LGTM, just a minor comment on where to create the guard.

**Pull request closed without merging**